feat: Add support for transactional table create to ConnectorMetadata

hdikeman opened the pull request:
Summary:
createTable is the interface used to essentially "start" the process of table creation. It's also implicit that this API must generate and stash a table in the ConnectorMetadata such that a Table/TableLayout is made available for planning via findTable
For certain (transactional) backends, it may not be possible to create and store the TableLayout before the write completes. For example, for a distributed Hive backend, the table is not typically created in the backend until TableFinish
In this change, doing the following things:

1. switching createTable to return a Table object directly without requiring (or allowing) separate lookup by caller in findTable
2. adding a separate write classification type kCreate, which covers bare CREATE TABLE and CTAS queries
3. removing the duplicate rowType present in beginWrite, and instead retrieving from the TableLayout

This API design is inspired by Presto, which has several separate begin/finish APIs for each particular modifying operation (create, insert, update, delete). One difference is that Presto generates a TableFinishNode during planning time which captures some write details (such as the table name, operation type, etc.) in what is effectively a write completion lambda. Since there is no physical table finish for Velox, we don't have a location in the plan to store the eventual table layout object, so we return it to the caller to capture how they will during optimization/execution rather than doing so in the ConnectorMetadata

I updated LocalHive so that it no longer retains the table being written until finishWrite is invoked

I updated the implementations of ConnectorMetadata to give the write-specific ConnectorMetadata APIs (createTable, beginWrite, finishWrite, rowHandles) UNSUPPORTED implementations in the ConnectorMetadata interface. It's expected many connectors will be read-only and will never implement write APIs, so this saves boilerplate from the majority of connectors which do not support write

Differential Revision: D83114656

MBkkt commented:
@hdikeman Hi, I have few questions, can you explain for me, please?

1) from current API and implementation state it's unclear to me why connector session needed as parameter? I think user can pass same variable inside insert table handle.

2) why does insert table handle named "insert" instead of "write"?

MBkkt reviewed (approved):
Looks better (then it was) to me, thanks!

MBkkt commented on axiom/connectors/ConnectorMetadata.h:
```diff
@@ -536,73 +540,38 @@ class ConnectorMetadata {
-      const TableLayout& layout,
+      const std::string& tableName,
       const velox::connector::ConnectorInsertTableHandlePtr& handle,
       const std::vector<velox::RowVectorPtr>& writerResult,
```
Shouldn't be this outside of connector metadata API?

I thought velox TableWriteMergeNode can be used for such logic

MBkkt commented on axiom/connectors/ConnectorMetadata.h:
```diff
@@ -536,73 +540,38 @@ class ConnectorMetadata {
 
+  /// Returns specification for repartitioning data before the table writer
+  /// stage.
+  virtual WritePartitionInfo writePartitionInfo(
```
In more later PR, @oerling removed it, although this removal didn't merged yet

MBkkt commented on axiom/connectors/ConnectorMetadata.h:
```diff
@@ -536,73 +540,38 @@ class ConnectorMetadata {
   /// finished. The result sets from the table writer fragments are passed as
   /// 'writerResults'. Their format and meaning is connector specific. the
   /// RowType is given by the outputType() of the TableWriter.
   virtual void finishWrite(
```
Interesting, @oerling changed this in his PR https://github.com/facebookexperimental/verax/pull/238 `bool success`

hdikeman commented:
Some additional discussion on this diff, posting here

Message 1, from Satadru
```
To Summarize, these are the two options.
Option1: No createTable at all, Henry already sent a diff for this: https://github.com/facebookexperimental/verax/pull/439 
Verax does not have access to Layout object from Connector
findTable for table returns nullptr until after finishWrite completes
virtual velox::connector::ConnectorInsertTableHandlePtr beginWrite(
  const std::string& tableName,
  const velox::RowTypePtr& rowType,
  const folly::F14FastMap<std::string, std::string>& options,
  WriteKind kind,
  const ConnectorSessionPtr& session) = 0;

virtual void finishWrite(
  const std::string& tableName,
  const velox::connector::ConnectorInsertTableHandlePtr& handle,
  const std::vector<velox::RowVectorPtr>& writerResult,
  const ConnectorSessionPtr& session) = 0;

Option2: createTable creates Table object but does NOT store it. If we all agree on this, we can switch to this option.
Verax retains Table from createTable, passes to caller of finishWrite
findTable for table returns nullptr until after finishWrite completes
virtual TablePtr createTable(
  const std::string& tableName,
  const velox::RowTypePtr& rowType,
  const folly::F14FastMap<std::string, std::string>& options,
  const ConnectorSessionPtr& session,
  TableKind tableKind = TableKind::kTable) = 0;

virtual velox::connector::ConnectorInsertTableHandlePtr beginWrite(
  const TableLayout& layout,
  const velox::RowTypePtr& rowType,
  const folly::F14FastMap<std::string, std::string>& options,
  WriteKind kind,
  const ConnectorSessionPtr& session) = 0;

virtual void finishWrite(
  const TableLayout& layout,
  const velox::connector::ConnectorInsertTableHandlePtr& handle,
  const std::vector<velox::RowVectorPtr>& writerResult,
  const ConnectorSessionPtr& session) = 0;
```

Message 2, from @oerling 
```
The only must have is that the Optimization knows the layout, partitioning, buckets, possible secondary indices etc. But this is satisfied by #2 so go ahead. Suggest passing the non-Metastore schema in ConnectorSession as there is already an argument for that in the signatures.
```

I removed rowType from beginWrite as well, since I believe this is the same RowType captured in the layout. Waiting for comment from @oerling on whether that makes sense, and also clarification on the ask "Suggest passing the non-Metastore schema in ConnectorSession as there is already an argument for that in the signatures."

MBkkt commented:
@hdikeman I also have thought about "how does async table creation/transaction commit supposed to be implemented with this interface?" WDYT?

MBkkt commented:
@hdikeman 

> Do you mean in terms of a multi-query transaction? Or the atomicity of a table write operation for a single query? 

No I mean, what to do if commit is async (in terms of C++) operation that returns future/coroutine-handle/etc, commit will be called inside `finishWrite`, how we suppose to do this?


> I did notice there is no API currently to generate a ConnectorSession object. Think we will need to add one when the optimizer portion of the write changes is added (not sure if that was in your or Orri's pending diffs)

I thought idea of current code is that user should create ConnectorSession somehow and then pass it to optimizer.
So the optimizer code shouldn't create ConnectorSession from what I see.

hdikeman commented:
@MBkkt 

> No I mean, what to do if commit is async (in terms of C++) operation that returns future/coroutine-handle/etc, commit will be called inside `finishWrite`, how we suppose to do this?

I suppose in that case, the future/coroutine would have to be waited for synchronously inside `finishWrite`. I was operating under the assumption that `finishWrite` is the last operation performed during query execution, so the commit must complete before it exits, or throw an error if it fails

In other words, I think even if `finishWrite` did return a future, the caller (Runner) would have to immediately wait for the future regardless before completing query execution, so it makes sense to keep those details internal to the ConnectorMetadata implementation

Did you have something else in mind?

> I thought idea of current code is that user should create ConnectorSession somehow and then pass it to optimizer. So the optimizer code shouldn't create ConnectorSession from what I see.

Makes sense. I had noticed it was connector-specific (HiveConnectorSession for example) so I had assumed the optimizer would be on the hook for generating the session. Or else we would need to pass a session for each type of connector, right?

Or if only using the ConnectorSession for writing (and assuming only one Connector is receiving writes for any given query plan), the caller of the Optimizer would need to know ahead of time which connector is being written to so it can create the right one

I don't have any strong opinion here, I think in any case my original statement holds that the Session would hold multi-query transactional details and perform cleanup for uncommitted changes if needed

hdikeman commented:
> > In other words, I think even if finishWrite did return a future, the caller (Runner) would have to immediately wait for the future regardless before completing query execution, so it makes sense to keep those details internal to the ConnectorMetadata implementation
> > Did you have something else in mind?
> 
> Yes, we have own runner and in our case whole task won't be blocking. It's postgres(tcp) connection that is responsible for this query, this is "coroutine". So it can handle such event as async commit in non-blocking way. Can we make this API async? Is it make sense for you?

Yes this makes sense to me, didn't consider that use-case. Do you have a preference on the particular return type? folly::Future<folly::Unit>?

hdikeman commented:
> > Do you have a preference on the particular return type?
> 
> I think maybe API like velox Task (ContinueFuture*)? (where Future isn't created if function sync). https://github.com/facebookincubator/velox/blob/main/velox/exec/Task.h#L232

I think maybe that one returns by reference parameter since the return type is already taken by the RowVector results

Perhaps we can just return it directly, convention being that an already satisfied future can be returned if the function is synchronous. Thoughts @MBkkt ?

We can also return by reference if you prefer that for whatever reason

hdikeman commented:
> > I did notice there is no API currently to generate a ConnectorSession object.
> 
> @hdikeman Interesting... then how do you use the APIs that require ConnectorSession?

@mbasmanova pass a default nullptr to those. ConnectorSession is implementation-specific so Verax does not look inside

hdikeman commented on axiom/connectors/ConnectorMetadata.h:
```diff
@@ -603,6 +585,11 @@ class ConnectorMetadata {
 
+  /// Returns specification for repartitioning data before the table writer
+  /// stage.
+  virtual WritePartitionInfo writePartitionInfo(
```
Sure I'll take it out. @MBkkt also pointed out to me that Orri removed this permanently (and moved the same functionality to the TableLayout). So should be fine to remove

MBkkt commented:
@hdikeman 

> I think maybe that one returns by reference parameter since the return type is already taken by the RowVector results

Ok, sounds good to me

hdikeman commented on axiom/connectors/hive/HiveConnectorMetadata.h:
```diff
@@ -165,6 +164,14 @@ class HiveConnectorMetadata : public ConnectorMetadata {
   virtual void validateOptions(
       const folly::F14FastMap<std::string, std::string>& options) const;
 
+  virtual velox::connector::ConnectorInsertTableHandlePtr
```
I realize it's a pretty thin wrapper now, I can probably consolidate it around the beginWrite interface itself

hdikeman commented on axiom/connectors/hive/tests/HiveConnectorMetadataTest.cpp:
```diff
@@ -133,11 +133,47 @@ TEST_F(HiveConnectorMetadataTest, createTable) {
+  auto table = metadata->createTable("test", tableType, options, session);
+  auto insertHandle = metadata->beginWrite(
+      *table->layouts()[0], options, WriteKind::kCreate, session);
+  auto handle = std::make_shared<core::InsertTableHandle>(
```
Sure

mbasmanova commented on axiom/connectors/hive/HiveConnectorMetadata.h:
```diff
@@ -122,20 +122,19 @@ class HiveConnectorMetadata : public ConnectorMetadata {
 
-  velox::connector::ConnectorInsertTableHandlePtr createInsertTableHandle(
-      const TableLayout& layout,
+  TablePtr createTable(
```
> Yes it adds quite a lot of boilerplate though.
 
 What kind of boilerplate do you refer to?

> it makes backwards compatibility harder

Backwards compatibility is hard regardless.

hdikeman replied:
>What kind of boilerplate do you refer to?

All the ConnectorMetadata implementations tend to have a lot of functions which return UNSUPPORTED. Esp for write support, many backends will never support write (e.g. TPCH). Bucketing and transaction support may be similar in that sense

I won't die on this hill, probably it will be less of an issue as APIs stabilize. Just something I noticed

mbasmanova replied:
I get it that TPC-H doesn't support write, but are there any other connectors that shouldn't support write. It could be that we created many connector too soon before the APIs were ready.

mbasmanova replied:
Are you making a point that many connectors are read-only? That would make sense. If so, we are saying that we intentionally want to make write APIs optional. That would be fine as well. If that's the case, let's explain this clearly in the PR description and provide not-supported implementation in the base class (ConnectorMetadata), then remove these from derived classes. Does that make sense?

hdikeman replied:
>Are you making a point that many connectors are read-only?

Yes. There may be other APIs in the future which many/most connectors will never implement as well

>If that's the case, let's explain this clearly in the PR description and provide not-supported implementation in the base class (ConnectorMetadata), then remove these from derived classes. Does that make sense?

Makes sense to me. Thanks for the discussion

hdikeman commented:
> > @mbasmanova pass a default nullptr to those. ConnectorSession is implementation-specific so Verax does not look inside
> 
> Hmm... I see. Then we may need to document that session can be nullptr.

I clarified this in the APIs which take a session (createTable, beginWrite, finishWrite)

mbasmanova commented on axiom/connectors/hive/tests/HiveConnectorMetadataTest.cpp:
```diff
@@ -133,11 +133,47 @@ TEST_F(HiveConnectorMetadataTest, createTable) {
+      tableType->names(),
+      std::nullopt,
+      handle,
+      false,
```
nit: `/*hasPartitioningScheme=*/false`

mbasmanova commented on axiom/connectors/hive/tests/HiveConnectorMetadataTest.cpp:
```diff
@@ -133,11 +133,47 @@ TEST_F(HiveConnectorMetadataTest, createTable) {
+      idGenerator->next(),
+      builder.planNode()->outputType(),
+      tableType->names(),
+      std::nullopt,
```
nit: `/*columnStatsSpec=*/columnStatsSpec`

mbasmanova commented on axiom/connectors/hive/HiveConnectorMetadata.h:
```diff
@@ -122,24 +122,21 @@ class HiveConnectorMetadata : public ConnectorMetadata {
+      WriteKind kind,
+      const ConnectorSessionPtr& session) override;
+
+  velox::ContinueFuture finishWrite(
```
We no longer need this, do we?

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+ * limitations under the License.
+ */
+
+#include "axiom/connectors/hive/LocalHiveConnectorMetadata.h"
```
Have you used hg mv to move the file? Is it just showing up as delete + add instead of a move?

hdikeman replied:
`hg blame` is still ok from my side, I can't send it here because it's internal but mercurial detected the move so should be ok. I'll use `hg mv` directly in the future

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+        ROW({"numWrittenRows", "fragment", "tableCommitContext"},
+            {BIGINT(), VARBINARY(), VARBINARY()});
+
+    auto idGenerator = std::make_shared<core::PlanNodeIdGenerator>();
```
not needed

hdikeman replied:
Which part is not needed?

This was inherited from the existing test, I have never used the DistributedPlanBuilder before, but it looks like it demands a shared pointer to a ID generator:
```
 29   /// Constructs a top level DistributedPlanBuilder.                                                               
 30   DistributedPlanBuilder(                                                                                          
 31       const runner::MultiFragmentPlan::Options& options,                                                           
 32       std::shared_ptr<velox::core::PlanNodeIdGenerator> planNodeIdGenerator,                                       
 33       velox::memory::MemoryPool* pool = nullptr);
```

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+    auto idGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+    auto builder = exec::test::PlanBuilder(idGenerator).values({values});
+
+    auto plan = std::make_shared<core::TableWriteNode>(
```
Any particular reason not to use PlanBuilder::tableWrite API?

hdikeman replied:
No reason, I did not rewrite that portion of the test, just added some stronger checks and pulled out the shared portion

I can try it to see if it'll work

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+  static constexpr int32_t kNumVectors = 5;
+  static constexpr int32_t kRowsPerVector = 10000;
+
+  void writeDataToTable(const TablePtr& table, const RowVectorPtr& values) {
```
writeDataToTable -> writeToTable

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+        .get();
+  }
+
+  bool compareColumnVectors(
```
not used?

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+  }
+
+  void compareTableLayout(
+      const connector::hive::HiveTableLayout* expected,
```
pass by reference to ensure it is not null

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+  void readAndCompareTableData(
+      const std::string& tableName,
+      const RowVectorPtr& expectedData) {
+    auto metadata = dynamic_cast<connector::hive::HiveConnectorMetadata*>(
+        ConnectorMetadata::metadata(velox::exec::test::kHiveConnectorId));
+    ASSERT_TRUE(metadata != nullptr);
```
This code is repeated multiple times? Perhaps, add a helper method or just store this in a member variable.

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+    }
+
+    auto idGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+    runner::test::DistributedPlanBuilder rootBuilder(
```
Do we need a DistributedPlanBuilder?

hdikeman replied:
Ok will probably have to be changed back when we add partitioning perhaps

I'll change it to a normal plan builder for now

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+    runner::test::DistributedPlanBuilder rootBuilder(
+        runnerOptions, idGenerator, pool_.get());
+    rootBuilder.tableScan(
+        tableName, tableType, {}, {}, "", tableType, assignments);
```
> {}, {}, ""

Would you add parameter names for these?

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+  readAndCompareTableData("test", data);
+}
+
+TEST_F(LocalHiveConnectorMetadataTest, createTableEmpty) {
```
createTableEmpty -> createEmptyTable

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+      table->layouts()[0]);
+  ASSERT_TRUE(expected != nullptr);
+
+  EXPECT_EQ(expected->partitionColumns().size(), 1);
+  EXPECT_EQ(expected->partitionColumns()[0], expected->columns()[0]);
+  EXPECT_EQ(expected->numBuckets().value(), 4);
+  EXPECT_EQ(expected->orderColumns().size(), 2);
+  EXPECT_EQ(expected->orderColumns()[0], expected->columns()[0]);
+  EXPECT_EQ(expected->orderColumns()[1], expected->columns()[1]);
+  EXPECT_EQ(expected->hivePartitionColumns().size(), 1);
+  EXPECT_EQ(expected->hivePartitionColumns()[0], expected->columns()[3]);
+  EXPECT_EQ(expected->fileFormat(), dwio::common::toFileFormat("parquet"));
+  EXPECT_EQ(expected->table().options().at("compression_kind"), "snappy");
```
Can we avoid copy-pasting so much code?

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,357 @@
+       {"ds", VARCHAR()}});
+
+  folly::F14FastMap<std::string, std::string> options = {
+      {"bucketed_by", "key1"},
+      {"sorted_by", "key1, key2"},
```
We would need to list the directory content and (1) make sure # of files matches # of buckets; (2) file names should follow naming pattern that uses bucket number; (3) read each file and verify that all rows indeed belong to the right bucket.

hdikeman replied:
Hmm tricky. How do we read a file directly into a RowVectorPtr?

LocalHiveConnectorMetadata also doesn't expose its data publicly, but I think I can make a protected method public to do that

hdikeman commented on axiom/connectors/hive/HiveConnectorMetadata.cpp:
```diff
@@ -135,13 +136,13 @@ HiveConnectorMetadata::createInsertTableHandle(
       compressionKind = velox::common::stringToCompressionKind(it->second);
     } else {
-      compressionKind = velox::common::CompressionKind::CompressionKind_ZSTD;
+      compressionKind = velox::common::CompressionKind::CompressionKind_NONE;
```
Can remove this and open a separate one after this lands if you like, and add a test to enforce the createTable defaults for Hive

Alternatively I could pass no options to one of the createTable tests and check the defaults there, and keep them combined

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+  static constexpr int32_t kRowsPerVector = 10000;
+
+  std::shared_ptr<velox::connector::hive::HiveConnectorSplit>
+  makeHiveConnectorSplit(
```
Can you use one of the methods from the base class?

See velox/exec/tests/utils/HiveConnectorTestBase.h

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+    auto result = exec::test::AssertQueryBuilder(plan).copyResults(pool());
+    metadata_
+        ->finishWrite(
+            *table->layouts()[0],
```
Let's create a temp variable to hold table->layouts()[0] to avoid copy-pasting.

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+      const std::vector<std::string>& files,
+      const std::unordered_map<std::string, std::string>& partitions,
+      dwio::common::FileFormat format) {
+    auto table = metadata_->findTable(tableName);
+    auto& layouts = table->layouts();
+    EXPECT_EQ(1, layouts.size());
+    auto* layout =
+        dynamic_cast<const connector::hive::HiveTableLayout*>(layouts[0]);
+    EXPECT_TRUE(layout != nullptr);
```
This code repeats multiple times. Perhaps, introduce a helper method to take tableName and return the layout.

MBkkt commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+            *table->layouts()[0],
+            insertHandle,
+            {result},
+            WriteKind::kCreate,
```
For same reason I think we don't need to pass table layout here, because this information already contains in first `beginWrite` call

https://github.com/facebookincubator/axiom/pull/439#discussion_r2383296435

MBkkt replied:
@mbasmanova 

### Question 1

Can you explain what is use-case for `createTable` except `select into/etc` ("create table as select (CTAS, runs the query and saves results into a new table")?

I see you wrote about this empty table, but for an example later maybe someone wanted to remove table, or create index, or add column or something else.

### Question 2

> for each layout:
    beginLayout(table, layout, writeKind, options, session) -> handle
    finishLayout(handle, veloxWriteResult, session) -> finishResult
    finish(<list of finishResult>, session)

So it will be such API?

```cpp
TablePtr createTable(
      const std::string& tableName,
      const velox::RowTypePtr& rowType,
      const folly::F14FastMap<std::string, std::string>& options,
      const ConnectorSessionPtr& session)

velox::connector::ConnectorInsertTableHandlePtr beginWrite(
      const TablePtr& table,
      const TableLayout& layout,
      const folly::F14FastMap<std::string, std::string>& options,
      WriteKind kind,
      const ConnectorSessionPtr& session);

velox::ContinueFuture finishWrite(
      const velox::connector::ConnectorInsertTableHandlePtr& handle,
      const std::vector<velox::RowVectorPtr>& writerResult,
      const ConnectorSessionPtr& session);
```

If it's true, that exactly what I expected for `beginWrite` and `finishWrite`. In a few messages early I wrote about we don't need to pass all these parameters to `finishWrite` and we need to pass `table` in `beginWrite`.


### Question 3

Maybe we can avoid DDL only workload in Axiom?
I think Velox and Axiom API isn't ready for this.

Because there's a lot of different ways to do this and different DDL statements.
For an example drop ddl, create index, add columns, etc. 

From what I see in our usecase it can be easily separated and it's quite different.

The only usecase that we have where DDL mixed with DQL/DML is `select into`/`create table as`

### Question 4

How do you feel about remove `createTable` API from connector?

Instead we can say that user needs to provide `TablePtr` if this table is the one that created for this query (`select into`).

For an example in `PlanBuilder`

`tableWrite(connector_name, table_name, ...)` => will use schema resolver for lookup table for write
`tableWrite(connector_table_ptr, ...)` => will use provided table for write

MBkkt replied:
I'm thinking about something like this

```
user:
create table

optimizer:
list of handles
for each layout
  handles.push(beginWrite(...));

runner:
finishWrite(handles, results, true/false);
```

hdikeman replied:
Thanks for the discussion @mbasmanova and @MBkkt, sorry to join this late. Few thoughts:

1. if abortWrite is optional/best effort, does it make sense to include the API? e.g. if a connector must guard against the case where this API is not called, and provide a cleanup triggered by (WriteHandle?) GC, it seems preferable not to include an abort

2. As for the proposal on what additional information to include in the WriteHandlePtr:

>RowType describing the output of TableWriteNode
partitionColumns
partitionType (from https://github.com/facebookincubator/axiom/pull/238)
per-column statistics to collect (a.k.a columnStatsSpec in TableWriteNode)

This implies a requirement "you should be able to construct a TableWriteNode given the WriteHandlePtr", instead "You should be able to commit a write operation given the WriteHandlePtr and results from a TableWrite/TableWriteMerge node" makes more sense as a requirement

It seems most connectors will derive the information about the above from the WriterResults, or may stash it in their implementation-specific WriteHandle if not

3. As for this point:

>First, we agreed to not worry about supporting connectors that require the data to be written multiple types using multiple layouts

So far, a connector of this type doesn't exist, nor does the optimizer use any but the first layout. Is there a roadmap for using this functionality?

If we are likely not to, perhaps this makes sense to factor out of the code. I'm not sure I understand the use-case, @mbasmanova perhaps you understand the intention there better

hdikeman replied:
I updated with the new interfaces we agreed to above, thanks for the back and forth @mbasmanova and @MBkkt

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+
+  writeToTable(table, options, data, dwio::common::FileFormat::PARQUET);
+
+  auto* expected = dynamic_cast<const connector::hive::HiveTableLayout*>(
```
The layout is created in metadata_->createTable, hence, it is more logical to verify it right after that instead of after writeToTable.

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+  auto bucketFunction =
+      std::make_unique<velox::connector::hive::HivePartitionFunction>(
+          4, std::vector<velox::column_index_t>{0});
+  std::unordered_map<std::string, uint32_t> buckets = {
```
It might be cleaner to use std::vector where index == bucket (since buckets are sequential zero-based numbers)

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+    if (file.find(".schema") != std::string::npos) {
+      continue;
+    }
+    for (const auto& [bucket, index] : buckets) {
```
Can we avoid nested loop? I assume file names should end in 00000N_0. If so, we can simply extract last 8 characters and parse bucket number from these.

hdikeman replied:
It's actually not quite that easy since I think they also get a `.parquet` suffix

I guess I can search for `_0.parquet` and cast the previous 6 characters to an int. Probably can't look for just _0 since it might hit on the temp directory, I remember it being tmp_[RANDOM]

hdikeman replied:
It's actually like "/ds=2022-09-01/000000_0_TaskCursorQuery_0.parquet` but I just cut from the partition forwards since that's known

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+        partitions.resize(result->size());
+        bucketFunction->partition(*result, partitions);
+        for (auto i = 0; i < result->size(); ++i) {
+          EXPECT_EQ(partitions[i], index);
```
Thank you for adding this test.

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+  }
+  EXPECT_EQ(bucketFiles, 4);
+
+  compareTableLayout(*expected, "test");
```
Can we move this up and drop the following checks? These appear redundant.

```
  EXPECT_EQ(expected->partitionColumns().size(), 1);
  EXPECT_EQ(expected->partitionColumns()[0], expected->columns()[0]);
  EXPECT_EQ(expected->numBuckets().value(), 4);
  EXPECT_EQ(expected->orderColumns().size(), 2);
  EXPECT_EQ(expected->orderColumns()[0], expected->columns()[0]);
  EXPECT_EQ(expected->orderColumns()[1], expected->columns()[1]);
  EXPECT_EQ(expected->hivePartitionColumns().size(), 1);
  EXPECT_EQ(expected->hivePartitionColumns()[0], expected->columns()[3]);
  EXPECT_EQ(expected->fileFormat(), dwio::common::toFileFormat("parquet"));
  EXPECT_EQ(expected->table().options().at("compression_kind"), "zstd");
```

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+      metadata_->createTable("test_empty", tableType, options, session);
+
+  auto emptyData = makeRowVector(
+      tableType->names(),
```
makeRowVector(tableType, 0)

hdikeman commented on axiom/connectors/hive/LocalHiveConnectorMetadata.cpp:
```diff
@@ -879,13 +872,16 @@ void LocalHiveConnectorMetadata::createTable(
-    token = folly::trimWhitespace(token);
+  it = options.find("partitioned_by");
+  if (it != options.end()) {
+    folly::split(",", it->second, tokens);
+    for (auto& token : tokens) {
+      token = folly::trimWhitespace(token);
+    }
```
Not totally. Elsewhere we push right into JSON. Here we need to add some column name and type info

I can pull the JSON version of this into a helper though

hdikeman commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+      assignments[tableType->nameOf(i)] =
+          metadata_->createColumnHandle(*layout, tableType->nameOf(i));
+    }
+    auto plan = exec::test::PlanBuilder(pool())
```
>  88   /// The memory pool is used when parsing expressions containing complex-type
  89   /// literals, e.g. arrays, maps or structs. The memory pool can be empty if
  90   /// such expressions are not used in the plan.

Thanks wasn't aware of this

hdikeman commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+            .customizeData = customize1},
+    };
+
+    parquet::registerParquetReaderFactory();
```
Sure. I'll keep the writer though

hdikeman commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+       {"ds", VARCHAR()}});
+
+  folly::F14FastMap<std::string, std::string> options = {
+      {"bucketed_by", "key1"},
```
Where were you thinking

Also need to consider that it may not just be Hive using these constants. Other connectors may as well, since in the case of Presto these are the options supplied at the command line during a query `CREATE TABLE test WITH (partitioned_by = ARRAY['ds'], etc....) AS [...]`. Not all are connector-specific

mbasmanova replied:
I think we need to document what options are available. Perhaps, there are some generic options which would go into ConnectorMetadata.h. Other, connector-specific options would go into corresponding XxxConnectorMetadata.h. Would that work?

mbasmanova commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,411 @@
+
+    rowType_ = ROW({"c0"}, {BIGINT()});
+    testTables_ = {
+        runner::test::TableSpec{
```
Are you saying we'll remove TableSpec struct and related logic altogether? That's fine. Can you send a delete PR?

hdikeman replied:
I meant remove the usage from this test, LocalHiveConnectorMetadataTest. But sure

I do see one other usage of TableSpec in LocalRunnerTest, but that looks like it could be performed with a table write operation, and we can remove the TableSpec as you mentioned. Once the new API is committed I can replace those

hdikeman replied:
Resolving this for now since write changes have to land first before removing this test structure

mbasmanova commented on axiom/logical_plan/LogicalPlanNode.h:
```diff
@@ -660,21 +660,25 @@ using UnnestNodePtr = std::shared_ptr<const UnnestNode>;
-  kInsert = 1,
+  // A write operation to a new table which does not yet exist in the connector.
+  // Covers both creation of an empty table and create as select operations.
+  kCreate = 1,
```
I don't think we need changes to LogicalPlanNode in this PR. Let's remove these.

mbasmanova commented on axiom/connectors/ConnectorMetadata.h:
```diff
@@ -417,21 +417,32 @@ struct LookupKeys {
+
+  virtual ~ConnectorWriteHandle() = default;
+
+  const velox::connector::ConnectorInsertTableHandlePtr& insertHandle() const {
```
nit: perhaps, veloxHandle

mbasmanova commented on axiom/connectors/hive/LocalHiveConnectorMetadata.cpp:
```diff
@@ -804,38 +804,40 @@ void createDir(const std::string& path) {
   }
 }
+
+// Helper: split a set of string tokens with ',' delimiter
```
drop "Helper: "; this is redundant.

// Split a comma-separated string into tokens.

hdikeman replied:
Thanks, just copied the function comments from the two above. I can remove "Helper:" from those too

mbasmanova reviewed (commented):
@hdikeman Looks great to me % a few small nits.

hdikeman commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,374 @@
+    std::string outputPath = metadata_->tablePath(table->name());
+    auto session = std::make_shared<connector::hive::HiveConnectorSession>();
+    auto handle =
+        metadata_->beginWrite(table, options, WriteKind::kCreate, session);
```
Perhaps I can add this to the empty table test, after the table is created

hdikeman commented on axiom/connectors/hive/LocalHiveConnectorMetadata.cpp:
```diff
@@ -912,22 +909,30 @@ void LocalHiveConnectorMetadata::createTable(
-void LocalHiveConnectorMetadata::finishWrite(
-    const TableLayout& layout,
-    const velox::connector::ConnectorInsertTableHandlePtr& handle,
+velox::ContinueFuture LocalHiveConnectorMetadata::finishWrite(
```
Above is what I did for the time being

mbasmanova replied:
I see. Let's document this limitation in the PR description and comments for the LocalHiveConnectorMetadata APIs.

hdikeman commented:
Did the following in most recent update:

1. added abortWrite for Local Hive (only for CREATE TABLE or CREATE TABLE AS SELECT though)
2. added a test for abort and a test for insert
3. removed LogicalPlan changes
4. few updates to internal test APIs

hdikeman commented on axiom/connectors/hive/tests/LocalHiveConnectorMetadataTest.cpp:
```diff
@@ -0,0 +1,435 @@
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/connectors/hive/HiveConnectorSplit.h"
+#include "velox/connectors/hive/HivePartitionFunction.h"
+#include "velox/dwio/parquet/RegisterParquetReader.h"
```
Thanks yeah forgot to remove that when I dropped the duplicate reader registration, thanks

hdikeman commented:
There was a new invocation of `TestConnector->createTable` I had to fix, since I rebased on latest master

Should be good now

mbasmanova commented on axiom/connectors/hive/HiveConnectorMetadata.h:
```diff
@@ -98,6 +98,55 @@ class HiveTableLayout : public TableLayout {
+  /// to individual buckets.
+  static constexpr auto kSortedBy = "sorted_by";
+
+  /// The table storage format. See velox::dwio::common::FileFormat.
```
This is an optional property, right? Would you document the default value? Ditto for other properties.

mbasmanova commented on axiom/connectors/hive/HiveConnectorMetadata.h:
```diff
@@ -98,6 +98,55 @@ class HiveTableLayout : public TableLayout {
+  /// Comma-delimited list of columns to bucket the table by.
+  static constexpr auto kBucketedBy = "bucketed_by";
+
+  /// The number of buckets to create the table with.
```
Would you clarify that both bucketed_by and bucket_count must be specified or none? Are there restrictions on bucket_count? Does it have to be > 0 and power of 2?

mbasmanova replied:
Let's add power-of-2 requirement and enforcement. We can relax it later if there is ever a need. Relaxing is easier than tightening.

mbasmanova reviewed (approved):
Thank you!
